Filter resources by topic

alinosratipour opened the pull request:
This is a:

<!-- Tick one category - if more than one applies, it should be split up -->

- [x] ✨ **New feature** - new behaviour has been implemented
- [ ] 🐛 **Bug fix** - existing behaviour has been made to behave
- [ ] ♻️ **Refactor** - the behaviour has not changed, just the implementation
- [ ] ✅ **Test backfill** - tests for existing behaviour were added but the behaviour itself hasn't changed
- [ ] ⚙️ **Chore** - maintenance task, behaviour and implementation haven't changed

<!-- adapted from https://gitmoji.dev/ -->

### Description

<!-- Describe what merging this pull request will do -->

- **Purpose** - <!-- Allow astronauts to perform a case-insensitive search for their spaceship. -->
 we want user to be able to filter resource based on Topic we also want to make sure filter is apply to all pages not just current pages which means if we have same topic in other pages as well they must be shown when filter topic apply.
<!-- Describe how the reviewer should check it works -->

- **How to check** - <!-- Log in as an astronaut, go to the Spaceships tab and type "saturn" into the search box. Previously Saturn V would not have appeared, due to the capital S, but now it does. -->

You can check out to FilterTopic branch and `npm run dev`  choose topic from drop down menu and see the result or see screen recording below
[Screencast from 11-03-24 12:02:33.webm](https://github.com/CodeYourFuture/tech-products-demo/assets/74776647/598c70fa-f373-42e1-8692-3dd2abc09f81)



### Links

- CodeYourFuture/tech-products-demo#30
<!-- links to other issues/PRs/tickets, e.g. user/repo#123 -->

### Author checklist

<!-- All PRs -->

- [x] I have written a title that reflects the relevant ticket
- [x] I have written a description that says what the PR does and how to validate it
- [x] I have linked to the project board ticket (and any related PRs/issues) in the Links section
- [ ] I have added a link to this PR to the ticket
- [ ] I have made the PR to `main` from a branch named `<category>/<name>`, e.g. `feature/edit-spaceships` or `bugfix/restore-oxygen`
- [x] I have manually tested that the app still works correctly
- [x] I have requested reviewers here and in my team chat channel
<!-- depending on the task, the following may be optional -->
- [ ] I have spoken with my PM or TL about any parts of this task that may have become out-of-scope, or any additional improvements that I now realise may benefit my project
- [x] I have added tests, or new tests were not required
- [ ] I have updated any documentation (e.g. diagrams, schemas), or documentation updates were not required

textbook commented on client/src/components/ResourceList/index.js:
```diff
@@ -1,29 +1,101 @@
+			try {
+				const allResources = await resourceService.getPublished({});
+				const filtered = allResources.resources.filter(
+					(topic) => topic.topic_name === selectedTopic
```
❌ What you've named `topic` here is actually the _resource_, or you could just destructure it:
```suggestion
					({ topic_name }) => topic_name === selectedTopic
```

alinosratipour replied:
thank you @textbook that's done.

textbook commented on client/src/components/ResourceList/index.js:
```diff
@@ -1,29 +1,101 @@
+
+				setTopics(fetchedTopics);
+			} catch (error) {
+				throw new Error(`Error fetching topics: ${error.message}`);
```
💡  I'd generally avoid catching an error just to throw it again, I don't think adding a message that only appears in the dev console adds worthwhile context here

alinosratipour replied:
done thank you

textbook commented on client/src/components/ResourceList/index.js:
```diff
@@ -1,29 +1,101 @@
+		};
+
+		fetchTopics();
+	}, [setTopics, topicService]);
```
💡 `useState`'s setters are guaranteed stable references, I don't think you need to include them in dependency lists.

alinosratipour replied:
done

textbook commented on client/src/components/ResourceList/index.js:
```diff
@@ -1,29 +1,101 @@
+	}, [selectedTopic, resourceService, resources]);
+
+	const handleChange = (event) => {
+		const selectedValue = event.target.value;
+		const selectedOption = topics.find((option) => option.id === selectedValue);
+		setSelectedTopic(selectedOption ? selectedOption.name : "");
```
💡 This would be simpler if it was the _ID_ that was stored as `selectedTopic`; that would require enriching the `GET /resources` responses with the topic's ID as well as its name.

alinosratipour replied:
> 💡 This would be simpler if it was the _ID_ that was stored as `selectedTopic`; that would require enriching the `GET /resources` responses with the topic's ID as well as its name.

you are right thank you @textbook better to use ID but why we need to enrich GET/resource ? resources prop contains topic id we can extract it from there i changed it and it works.

textbook commented on client/src/components/ResourceList/index.js:
```diff
@@ -1,29 +1,101 @@
-					<em>No resources to show.</em>
-				</li>
+		<>
+			{resources.length > 0 && pathname !== "/drafts" && (
```
❓ I'm in two minds about the check on `resources.length` here. On the one hand yes there's no point letting the user filter no resources, but on the other hand we will happily show no resources for a given topic _after_ they've filtered the list! Might be simpler and more consistent to leave it out.

❌ I really don't like the `pathname` handling here. `!== "/drafts"` (rather than `=== "/"`) means that every time we add a new display of resources we have to explicitly opt _out_ of this behaviour, instead of opting _in_ when we actually wanted it.

💡 But if this is basically just a boolean toggle, "allow topic filter if `"/"`, don't if `"/drafts"`", why not simply let the parent _pass a boolean_, e.g. `allowTopicFiltering`? That also means we don't have to maintain a growing array e.g. `["/", "/foo", "/bar"].includes(pathname)` as we add more uses, which breaks the [open-closed principle](https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle).

alinosratipour replied:
> ❓ I'm in two minds about the check on `resources.length` here. On the one hand yes there's no point letting the user filter no resources, but on the other hand we will happily show no resources for a given topic _after_ they've filtered the list! Might be simpler and more consistent to leave it out.
> 
> ❌ I really don't like the `pathname` handling here. `!== "/drafts"` (rather than `=== "/"`) means that every time we add a new display of resources we have to explicitly opt _out_ of this behaviour, instead of opting _in_ when we actually wanted it.
> 
> 💡 But if this is basically just a boolean toggle, "allow topic filter if `"/"`, don't if `"/drafts"`", why not simply let the parent _pass a boolean_, e.g. `allowTopicFiltering`? That also means we don't have to maintain a growing array e.g. `["/", "/foo", "/bar"].includes(pathname)` as we add more uses, which breaks the [open-closed principle](https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle).

@textbook that's done

textbook commented on client/src/components/ResourceList/index.js:
```diff
@@ -1,29 +1,101 @@
-			))}
-		</ul>
+			<ul className="resource-list">
+				{filteredResources.length === 0 && (
```
💡 Using `filteredResources` down here is quite confusing, because sometimes they're **not** filtered (`setFilteredResources(resources)`).

alinosratipour replied:
> 💡 Using `filteredResources` down here is quite confusing, because sometimes they're **not** filtered (`setFilteredResources(resources)`).

@textbook  done tahnk you (:+1:

textbook commented on client/src/components/ResourceList/index.js:
```diff
@@ -1,29 +1,101 @@
 
-export default function ResourceList({ publish, resources }) {
+export default function ResourceList({ publish, resources, pathname }) {
+	const [topics, setTopics] = useState([]);
+	const topicService = useService(TopicService);
+	const [selectedTopic, setSelectedTopic] = useState("");
+	const [filteredResources, setFilteredResources] = useState([]);
+	const resourceService = useService(ResourceService);
```
💡 I'd have done:
```suggestion
	const [filteredResources, setFilteredResources] = useState([]);
	const [selectedTopic, setSelectedTopic] = useState("");
	const [topics, setTopics] = useState([]);
	const resourceService = useService(ResourceService);
	const topicService = useService(TopicService);
```

💡 the default for `filteredResources` might as well be `resources`

💡 maybe `undefined` would be a better default topic than `""`?

alinosratipour replied:
@textbook done thanks (:+1

textbook commented on client/src/components/ResourceList/index.js:
```diff
@@ -1,29 +1,101 @@
 import "./ResourceList.scss";
 
-export default function ResourceList({ publish, resources }) {
+export default function ResourceList({ publish, resources, pathname }) {
```
❌ This has gone from a fairly simple presentational component, driven entirely by two props, to a complex, stateful one. I can't see why that's necessary. A few things that occur to me:

- We call `resourceService.getPublished({})` again every time we need to "fetch resources by topic", but:
- It actually _doesn't_ fetch them by topic, just re-fetches the array we were already looking at (i.e. what's already in the prop named `resources`); then
- Filters them on the client side (guaranteeing we'll see _at most_ one page's worth, and most likely much less)

We could get the same behaviour with something like:

```javascript
const displayResources = useMemo(() => {
  return selectedTopic
    ? resources.filter(/* ... */)
    : resources;
}, [resources, selectedTopic]);
```

💡 alternatively this complexity could be hoisted out of this component entirely, and dealt with by the parent _providing_ the `resources` prop. That way we wouldn't need to toggle whether or not filtering should be allowed, it wouldn't be the responsibility of this component at all.

textbook replied:
Actually I realise it's not the same as `resources`, it's _always_ the first 20. So the suggestion I made above would be an improvement, it would interact more coherently with pagination.

alinosratipour replied:
@textbook  thank you for you valuable feedback I will soon address all above comments i push it again :+1:

alinosratipour replied:
done

textbook reviewed (changes requested):
## UX/UI

✅ First acceptance criterion passed, I can filter the list to show only resources of a given topic:

![Screenshot 2024-02-01 at 01 28 23](https://github.com/CodeYourFuture/tech-products-demo/assets/785939/2fe21c6a-7c53-4f37-989e-eddc32e33565)

❌ Second acceptance criterion failed, the pagination doesn't work correctly with topic filtering - it won't show resources from subsequent pages, and the number of pages shown doesn't correspond with the number of filtered resources

❓ It wasn't an explicit requirement, but did you realise you can't _remove_ filtering by topic?

❓ Would it make sense for this to be closer to the pagination controls?

## Implementation

✅ Resource list component uses topci service (mostly) sensibly

✅ Existing tests have been updated accordingly to keep pipeline passing

❌ Resource list has gained quite a lot of complexity but no additional low-level test cases for that

❌ No E2E test cases (except a brief check that the select is rendered at all)

---

💡 Also for you specifically we'd talked about SQL stuff, which _still_ isn't shown here - even if we don't get to fixing pagination check my suggestion about using the topic ID, that would involve some.

---

**Key**:
- ✅ approval, nothing needs doing before merge
- ❌ problem, needs fixing before merge
- ❓ question, needs answering before merge
- 💡 suggestion, can be followed or ignored

alinosratipour commented on .vscode/settings.json:
```diff
@@ -2,7 +2,7 @@
 		"source.fixAll.eslint": true
 	},
-	"editor.formatOnSave": false,
+	"editor.formatOnSave": true,
```
> 💡 Try not to include changes that aren't relevant to the story in PRs

@textbook sure sorry I missed to undo that for i push.

textbook commented:
A note on the conflicts: you can rebase this branch if you'd like to, but **you are not required to**. #33 changed the way the frontend is built and tested, which (among other things) requires a `.jsx` extension for React JSX files.

alinosratipour commented on client/src/components/ResourceList/index.js:
```diff
@@ -1,29 +1,101 @@
+	useEffect(() => {
+		const fetchTopics = async () => {
+			try {
+				const isTestEnvironment = process.env.NODE_ENV === "test";
+
+				const fetchedTopics = isTestEnvironment
+					? []
+					: await topicService.getTopics();
```
@textbook the reason i did that to fix this error i was facing Error: Error fetching topics: Cannot read properties of null (reading '_location') when running test and i had no idea why that was happening. today i done some debugging and realised that  issue might be related it to this part of code for some reason test does not like undifined or null but its happy with empty array []
```
async getTopics() {
		const res = await this.fetch(TopicService.ENDPOINT);
		if (res.ok) {
			return res.json();
		}
	}
	
```

so I change it to this 
```
async getTopics() {
		try {
			const res = await this.fetch(TopicService.ENDPOINT);
			if (res.ok) {
				return res.json();
			} else {
				return [];
			}
		} catch (error) {
			return [];
		}
	}
}

```

also change  test to this
it("resolves undefined if the request fails", async () => {
		server.use(
			rest.get("/api/topics", (req, res, ctx) => res(ctx.status(500)))
		);
		// await expect(service.getTopics()).resolves.toBeUndefined();
		const topics = await service.getTopics();
		expect(topics).toEqual([]);
	});
	
	and test is now passing so code in our component will be this
	```
useEffect(() => {
		const fetchTopics = async () => {
			try {
				const fetchedTopics = await topicService.getTopics();
				setTopics(fetchedTopics);
			} catch (error) {
				throw new Error(`Error fetching topics: ${error.message}`);
			}
		};

		fetchTopics();
	}, [setTopics, topicService]);
	
	

```

alinosratipour commented on client/src/pages/Home/Home.test.js:
```diff
@@ -25,9 +25,15 @@ describe("Home", () => {
 					})
 				);
+			}),
+			rest.get("/api/topics", (req, res, ctx) => {
+				const mockTopics = [
+					{ id: "1", name: "Topic 1" },
+					{ id: "2", name: "Topic 2" },
+				];
+				return res(ctx.json(mockTopics));
 			})
```
@textbook done thank you (:+1

alinosratipour commented on client/setupTests.js:
```diff
@@ -3,7 +3,6 @@ import { randomUUID } from "node:crypto";
 import "@testing-library/jest-dom";
 import "whatwg-fetch";
 import { setupServer } from "msw/node";
-
```
done

alinosratipour commented on client/src/components/ResourceList/ResourceList.scss:
```diff
@@ -65,3 +65,12 @@ ul.resource-list {
 	}
 }
+
+.custom-select {
+	padding: layout.$baseline;
+	width: 100%;
+	font: typography.$body-text;
+	background-color: white;
+	border: 1px solid palette.$light-grey;
+	border-radius: layout.$border-radius;
+}
```
@textbook change made thanks(:+1

alinosratipour commented on client/src/components/Form/Select.js:
```diff
@@ -9,6 +9,8 @@ export default function Select({
 	options,
 	placeholder,
 	required,
+	onChange = () => {},
+	className,
```
> 💡 This seems like a reasonable extension to the behaviour that's already provided, but then all of the styles duplicate what's already in `src/components/Form/Form.scss`. Did you consider extending _the `Form` itself_, instead? With that approach:
> 
> ```js
> 	<Form onChange={handleChange}>
> 		<FormControls.Select
> 			label="Filter Topic"
> 			name="topic"
> 			options={topics}
> 			placeholder="Select a topic"
> 		/>
> 	</Form>
> ```
> 
> gives you something consistent with what's already on the suggestions page, and you just have to suppress the button (perhaps the same way the reset button is already handled).
> 
> ![Screenshot 2024-02-01 at 00 26 01](https://private-user-images.githubusercontent.com/785939/301350862-a5446f3b-9a7b-4712-bde9-0e9d21481cd0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg2ODM1ODQsIm5iZiI6MTcwODY4MzI4NCwicGF0aCI6Ii83ODU5MzkvMzAxMzUwODYyLWE1NDQ2ZjNiLTlhN2ItNDcxMi1iZGU5LTBlOWQyMTQ4MWNkMC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjIzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIyM1QxMDE0NDRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03OWY4NTQ0Mzg2ZmUxYzM4MTE3MTdmNjcyMDk4NWMzMmU5MDU1ZDg4YjQ4M2UyNWRmOWQ4N2M2MDdmMDI3MjQ5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.N9_ydV6QMi6sGEsbbm8ArfFaWYgeddlFlZLgK55uO2k)

i am wondering why we attach button to  select menu in case we might want to reuse it some other places like the one i did in resource list. wouldn't it be better to have separate button component rather than have it bind with select menu than we can use it anywhere without having to suppress  button every time we want to use select menu i understand this might me out side of scope of this ticket but as a suggestion. please correct me if i am wrong.

alinosratipour replied:
> > I think we need to be able to use the select menu on its own without having to wrap it around a form.
> 
> In that case a third option would be to have a separate select for use _outside_ a form, leaving this one as the select for use _inside_ a form (the "namespacing" of `FormControls.Select` was intended to indicate this).
> 
> What made me think about making _the form_ configurable was that it keeps all of the styling and our handling of controls more consistent.

Or we can render buttons only where they are needed, rather than within the Form and have Form component  to provide a reusable form structure. in that case we don't need reintroduce new select.

alinosratipour commented:
> ❓ Would it make sense for this to be closer to the pagination controls?

i am not sure since we don't have design input to follow

alinosratipour commented:
@textbook I Removed Home.test since i moved fetching resources into its own hook and also placed Pagination into resourceList therefor not functionality left in home page.

textbook reviewed (commented):
Thank you for iterating on this! I think we can consider this complete as far as the Tech Products application process is concerned.

If I were reviewing this as an actual change to merge (or not) into the codebase, I'd still want to do further work. Fundamentally this can't work in a sensible way without having an endpoint that handles filtering **and** pagination, so you can ask for e.g. _"the third page of 20 resources in the React topic"_.

Also I think changing the filter selection currently resets the user's page size choice, which is a bit awkward (and would particularly conflict with the intent of #17).

I've left some comments for your information, but no further changes are needed for the application.

alinosratipour commented:
> Thank you for iterating on this! I think we can consider this complete as far as the Tech Products application process is concerned.
> 
> If I were reviewing this as an actual change to merge (or not) into the codebase, I'd still want to do further work. Fundamentally this can't work in a sensible way without having an endpoint that handles filtering **and** pagination, so you can ask for e.g. _"the third page of 20 resources in the React topic"_.
> 
> Also I think changing the filter selection currently resets the user's page size choice, which is a bit awkward (and would particularly conflict with the intent of #17).
> 
> I've left some comments for your information, but no further changes are needed for the application.

@textbook  Thank you for your comment. I do agree with you If i understand it correctly we need to send topic id some how to backend and filter resource in there therefor we do not need to bring all data into front end and do filtering and pagination there.(:+1:

alinosratipour commented:
> > Thank you for iterating on this! I think we can consider this complete as far as the Tech Products application process is concerned.
> > If I were reviewing this as an actual change to merge (or not) into the codebase, I'd still want to do further work. Fundamentally this can't work in a sensible way without having an endpoint that handles filtering **and** pagination, so you can ask for e.g. _"the third page of 20 resources in the React topic"_.
> > Also I think changing the filter selection currently resets the user's page size choice, which is a bit awkward (and would particularly conflict with the intent of #17).
> > I've left some comments for your information, but no further changes are needed for the application.
> 
> @textbook Thank you for your comment. I do agree with you If i understand it correctly we need to send topic id some how to backend and filter resource in there therefor we do not need to bring all data into front end and do filtering and pagination there.

alinosratipour commented on client/src/components/ResourceList/ResourceList.test.js:
```diff
@@ -1,47 +1,184 @@
+	})
+);
+
+jest.mock("../../hooks/useFetchPublishedResources", () => ({
+	useFetchPublishedResources: jest.fn(),
+}));
+
+jest.mock("../../hooks/useFetchTopics", () => ({
+	useFetchTopics: jest.fn(),
+}));
```
> Rather than mock the hooks, which don't actually have their own tests and therefore aren't covered _anywhere_, why not mock the requests as you did with `/api/topics` above (which isn't actually used due to the mocking of `useFetchTopics`)?

yes, that can be another more reliable option thank you.

alinosratipour commented on client/src/pages/Drafts/Drafts.test.js:
```diff
@@ -25,6 +35,46 @@ describe("Drafts", () => {
 		).resolves.toHaveTextContent("foo");
 	});
 
+	it("shows the resource", () => {
```
> We already render the `DraftList` component in the context of the `Draft` page in the other test - I'm not sure this adds value, but also this is probably the wrong place for it (all tests in `Drafts.test.js` should `render(<Drafts {...} />)`, if you `render(<DraftList {...} />)` to test it _separately_ that should also be in a separate _file_ `DraftList.test.js`.

:+1:

alinosratipour commented on client/src/pages/Drafts/Drafts.test.js:
```diff
@@ -17,6 +19,14 @@ describe("Drafts", () => {
 		server.use(
 			rest.get("/api/resources", (req, res, ctx) => {
 				return res(ctx.json({ resources: [resource] }));
+			}),
+
+			rest.get("/api/topics", (req, res, ctx) => {
+				const mockTopics = [
+					{ id: "1", name: "Topic 1" },
+					{ id: "2", name: "Topic 2" },
+				];
+				return res(ctx.json(mockTopics));
```
> Is this actually needed? We don't load topics on the drafts page, and the test passes just fine with it commented out.

you are right I forgot to remove that :+1:

alinosratipour commented on client/src/services/topicService.test.js:
```diff
@@ -21,6 +21,7 @@ describe("TopicService", () => {
 			rest.get("/api/topics", (req, res, ctx) => res(ctx.status(500)))
 		);
-		await expect(service.getTopics()).resolves.toBeUndefined();
+		const topics = await service.getTopics();
+		expect(topics).toEqual([]);
```
> await expect(service.getTopics()).resolves.toEqual([]);

yes, you are right. thanks:+1:

alinosratipour commented on server/resources/resourcesService.js:
```diff
@@ -23,12 +23,16 @@ export async function getAll({ draft = false }, { page = 1, perPage = 20 }) {
 		page,
 		perPage,
-		resources,
+		resources: pagedResult,
+		allResources: findAllResource,
```
> This makes no sense at all to me. The point of paginating the resources is to allow the site to remain performant as the amount of data grows. If every "page" of resources also actually includes _all of the resources_ as well, there was no point at all in paginating to being with. In fact it's _worse_ than never having paginated, because as many as `perPage` resources are included in the response _twice_.
> 
> Also note there's no point making yet another DB query to get the total count (line 29) when that's the length of the array of all resources (as you actually assert on in the integration test).

you are right  I've got you point we need to handle filtering in backend  using topic id then send send filtered data into front end :+1:

alinosratipour commented on e2e/integration/home.test.js:
```diff
@@ -30,3 +30,23 @@ it("supports users without public emails", () => {
 	cy.findByRole("heading", { level: 1 }).should("contain.text", "Resources");
 });
+
+it("Filters resources by topic", () => {
```
> resourcesFixture

how about oneFilteredResourcesFixture ?
